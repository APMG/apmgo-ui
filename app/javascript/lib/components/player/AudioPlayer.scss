@import '../../styles/helpers';

.player {
  height: $player-height-mobile;
  padding: 20px;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: $z-player;
  text-align: center;
  color: $color-white;
  background-color: $color-grey-dark;

  @include media($media-small-tablet) {
    height: $player-height-small-tablet;
  }

  @include media($media-tablet) {
    height: $player-height-tablet;
  }
}

.title {
  margin-bottom: 15px;
  font-size: 1.8rem;
  font-weight: 700;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;

  @include media($media-small-tablet) {
    font-size: 2.2rem;
    margin-bottom: 20px;
  }

  @include media($media-tablet) {
    font-size: 2.6rem;
    margin-bottom: 30px;
  }
}

.controls {
  margin-bottom: 10px;
}

.playPause {
  margin-bottom: 10px;
}

.playPauseBtn {
  height: 48px;
  width: 48px;
  margin: 0 60px;
  padding: 0 4px;
  border: 0;
  fill: $color-white;
  vertical-align: middle;
  background: transparent;
  transition: background-color $transition;

  &:focus {
    outline: 0;
    background-color: rgba($color-white, .1);
  }
}

.skipBtn {
  height: 35px;
  width: 35px;
  position: relative;
  background: transparent;
  border: 0;
  padding: 12px 10px 8px;
  font-size: 1.3rem;
  vertical-align: middle;
  fill: $color-white;
  transition: background-color $transition;

  &:focus {
    outline: 0;
    background-color: rgba($color-white, .1);
  }

  svg {
    pointer-events: none;
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
    width: 100%;
  }
}

.skipBack {
  composes: skipBtn;
}

.skipFwd {
  composes: skipBtn;
}

.progress {
  margin-bottom: 15px;

  @include media($media-small-tablet) {
    margin-bottom: 20px;
  }
}

.timestamps {
  @include clearfix();
  margin-bottom: 10px;
}

.currentTime {
  float: left;
}

.duration {
  float: right;
}

.scrubber {
  width: 100%;
}

.volume {
  display: none;
}
