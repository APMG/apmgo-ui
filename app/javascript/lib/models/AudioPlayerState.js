// @flow
type AudioPlayerStateParams = {
  currentTrackId?: number,
  paused?: boolean,
  muted?: boolean,
  currentTime?: number,
  updateAudioElementTime?: boolean, // only present when explicitly set, signals need to update the time on the audio element
  canPlay?: boolean,
  volume?: number,
  duration?: number,
  isDragging?: boolean,
  playWhenCan?: boolean
}

export default class AudioPlayerState {
  currentTrackId: number | typeof undefined
  paused: boolean
  muted: boolean
  currentTime: number
  updateAudioElementTime: boolean
  canPlay: boolean
  volume: number
  duration: number
  isDragging: boolean
  playWhenCan: boolean

  constructor (params: AudioPlayerStateParams) {
    this.currentTrackId = params.currentTrackId
    this.paused = !!params.paused
    this.muted = !!params.muted
    this.currentTime = params.currentTime || 0
    this.updateAudioElementTime = params.updateAudioElementTime || false
    this.canPlay = params.canPlay || false
    this.duration = params.duration || 0
    this.isDragging = params.isDragging || false
    this.playWhenCan = params.playWhenCan || false

    if (params.volume === 0) {
      this.volume = 0
    } else {
      this.volume = params.volume || 1
    }
  }

  setCurrentTrackId (track: number) {
    return this._make({
      currentTrackId: track
    })
  }

  setTime (time: number) {
    return this._make({currentTime: time})
  }

  // If chaining AudioPlayerState updates,
  // this method must be called last
  // or it will be overwritten
  // because its value is not preserved in _make({})
  instanceUpdatesAudioElementTime () {
    return this._make({updateAudioElementTime: true})
  }

  setCanPlay (canPlay: boolean) {
    return this._make({canPlay: canPlay})
  }

  play (currentTrackId?: number) {
    currentTrackId = currentTrackId || this.currentTrackId

    return this._make({
      paused: false,
      currentTrackId: currentTrackId
    })
  }

  pause () {
    return this._make({paused: true})
  }

  mute () {
    return this._make({muted: true})
  }

  unmute () {
    return this._make({muted: false})
  }

  setVolume (volume: number) {
    return this._make({volume: volume})
  }

  setDuration (duration: number) {
    return this._make({duration: duration})
  }

  setPlayWhenCan (playWhenCan: boolean) {
    return this._make({playWhenCan: playWhenCan})
  }

  _make (params: {}): AudioPlayerState {
    let newParams = {
      currentTrackId: this.currentTrackId,
      paused: this.paused,
      muted: this.muted,
      currentTime: this.currentTime,
      canPlay: this.canPlay,
      volume: this.volume,
      duration: this.duration,
      isDragging: this.isDragging,
      playWhenCan: this.playWhenCan,
      // note doUpdateTime is not here by default,
      // it must be set explicitly, meaning that if
      // methods are chained, it will be overwritten
      // if instanceUpdatesAudioElementTime() is not called last
      ...params
    }

    return new AudioPlayerState(newParams)
  }
}
