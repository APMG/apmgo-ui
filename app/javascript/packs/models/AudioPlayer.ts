import { AudioPlayerType } from '../redux/types'

export default class AudioPlayer implements AudioPlayerType {

  item_id: number
  paused: boolean

  constructor(item_id: number) {
    this.item_id = item_id
    this.paused = true;
  }
}