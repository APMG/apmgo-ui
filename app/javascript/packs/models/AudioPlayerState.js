// @flow

type AudioPlayerStateParams = {
  currentTrackId?: number,
  paused?: boolean,
  muted?: boolean,
  currentTime?: number
}

export default class AudioPlayerState {
  currentTrackId: number | typeof undefined
  paused: boolean
  muted: boolean
  currentTime: number

  constructor(params: AudioPlayerStateParams) {

    this.currentTrackId = params.currentTrackId
    this.paused = !!params.paused
    this.muted = !!params.muted
    this.currentTime = params.currentTime || 0
  }

  setCurrentTrackId(track: number) {
    return this._make({
      currentTrackId: track
     })
  }

  setTime(time: number) {
    return this._make({currentTime: time})
  }

  play(currentTrackId?: number) {
    let params : {paused: boolean, currentTrackId?: number} = { paused: false }

    if (currentTrackId) {
      params.currentTrackId = currentTrackId
    }

    return this._make(params)
  }

  pause() {
    return this._make({paused: true})
  }

  mute() {
    return this._make({muted: true})
  }

  unmute() {
    return this._make({muted: false})
  }

  _make(params: {}) : AudioPlayerState {
    let newParams = {
      currentTrackId: this.currentTrackId,
      paused: this.paused,
      muted: this.muted,
      currentTime: this.currentTime,
      ...params
    }
    return new AudioPlayerState(newParams)
  }
}
