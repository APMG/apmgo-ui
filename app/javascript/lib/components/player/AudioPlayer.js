// @flow
import React from 'react';
import { connect, dispatch } from 'react-redux'

import AudioPlayerState from '../../models/AudioPlayerState';
import type { PlaylistItemType } from '../../redux/types';
import PlayPauseButton from './PlayPauseButton'
import MuteButton from './MuteButton'
import TimeScrubber from './TimeScrubber'
import TimeKeeper from './TimeKeeper'
import VolumeSlider from './VolumeSlider'
import PlayTimeDisplay from './PlayTimeDisplay'

import {
  audioMetaDataLoaded,
  audioCanPlay,
  playClick,
  pauseClick,
  muteClick,
  unmuteClick,
  updatePlayTime,
  timeScrubberChange,
  volumeChange,
  changeTrack
} from '../../redux/audio-player'

import { archivePlaylistItem } from '../../redux/playlist'

type AudioPlayerProps = {
  item: PlaylistItemType,
  audioPlayer: AudioPlayerState,
  audioCanPlay: () => {},
  audioMetaDataLoaded: (duration: number) => {},
  pause: () => {},
  play: () => {},
  mute: () => {},
  unmute: () => {},
  updatePlayTimeTimeKeeper: () => {},
  updatePlayTimeTimeScrubber: () => {},
  timeScrubberChange: () => {},
  updateVolume: () => {},
  onEnded: () => {},
  changeTrack: (item: PlaylistItemType) => {}
}

export class AudioPlayerPresenter extends React.Component {

  audio: HTMLAudioElement
  props: AudioPlayerProps

  constructor(props: AudioPlayerProps) {
    super(props)
  }

  componentDidMount() {
    // by now, this.audio is an HTMLAudioElement ref
    // it gets set when the component renders...
    // see the audio element's `ref` attribute

    this.audio.src = this.props.item.attributes.audio_url
    this.audio.oncanplay = this.props.audioCanPlay,
    this.audio.onloadedmetadata = this.metaDataLoaded.bind(this)
    this.audio.onended = this.props.onEnded

    // So, these are being set here to make the component testable
    // the audio element does not get rendered in tests
    // so that logic needed to be moved out of the render function
    // to here, where it can be overridden by a test class
  }

  componentWillReceiveProps(newProps: AudioPlayerProps) {

    this._setPlayPaused(newProps.audioPlayer.paused)
    this.audio.muted = newProps.audioPlayer.muted
    this.audio.volume = newProps.audioPlayer.volume

    if(newProps.item.id !== this.props.item.id) {
      this.audio.src = newProps.item.attributes.audio_url
      if(this.audio.canPlay && !this.props.audioPlayer.paused) {
        this.audio.play()
      }
    }

    if(newProps.item.id !== newProps.audioPlayer.currentTrackId) {
      this.props.changeTrack(newProps.item)
    }

    if(newProps.audioPlayer.updateAudioElementTime) {
      this.audio.currentTime = newProps.audioPlayer.currentTime
    }
  }

  _setPlayPaused(paused: boolean) {
    if (paused !== this.audio.paused) {
      if (paused) {
        this.audio.pause()
      } else {
        this.audio.play()
      }
    }
  }

  metaDataLoaded(a:any,b:any,c:any) {
    this.audio.currentTime = this.props.item.attributes.playtime
    this.props.audioMetaDataLoaded(this.audio.duration)
  }

  audioRefCallback(ref:any) {
    if (!this.audio) {
      this.audio = ref
    }
  }

  render() {
    return (
      <div>
        <h3>
            { this.props.item
              ? 'Now playing: ' + this.props.item.attributes.audio_title
              : 'Loading ...' }
        </h3>

        <audio
          ref={ (ref) => this.audio=ref }>
        </audio>

        <PlayPauseButton
          paused={this.props.audioPlayer.paused}
          canPlay={this.props.audioPlayer.canPlay}
          pause={this.props.pause}
          play={this.props.play}
        />
        <MuteButton
          muted={this.props.audioPlayer.muted}
          mute={this.props.mute}
          unmute={this.props.unmute}
        />

        <h3>Time Control</h3>
        <TimeKeeper // this component is invisible
          audio={ this.audio }
          updatePlayTime={this.props.updatePlayTimeTimeKeeper}
        />
        <PlayTimeDisplay
          currentTime={this.props.audioPlayer.currentTime}
        />
        <TimeScrubber
          updatePlayTime={this.props.updatePlayTimeTimeScrubber}
          timeScrubberChange={this.props.timeScrubberChange}
          paused={this.props.audioPlayer.paused}
          duration={this.props.audioPlayer.duration}
          currentTime={this.props.audioPlayer.currentTime}
        />

        <h3>Volume Control</h3>
        <VolumeSlider
          volume={this.props.audioPlayer.volume}
          updateVolume={this.props.updateVolume}
        />
      </div>
    )
  }
}

export const mapDispatchToProps = (dispatch: dispatch, ownProps: AudioPlayerProps) => {
  return {
    audioCanPlay: () => {
      dispatch(audioCanPlay())
    },
    audioMetaDataLoaded: (duration: number) => {
      dispatch(audioMetaDataLoaded(duration))
    },
    play: () : void => {
      dispatch(playClick(ownProps.item.id))
    },
    pause: (): void => {
      dispatch(pauseClick(ownProps.item.id))
    },
    mute: () => {
      dispatch(muteClick())
    },
    unmute: () => {
      dispatch(unmuteClick())
    },
    updatePlayTimeTimeKeeper: (currentTime: number) => {
      dispatch(updatePlayTime(ownProps.item.id, Math.ceil(currentTime)))
    },
    updatePlayTimeTimeScrubber: (currentTime: number) => {
      dispatch(updatePlayTime(ownProps.item.id, Math.ceil(currentTime)))
    },
    timeScrubberChange: (currentTime: number) => {
      dispatch(timeScrubberChange(ownProps.item.id, Math.ceil(currentTime)))
    },
    updateVolume: (volume: number) => {
      dispatch(volumeChange(volume))
    },
    onEnded: () => {
      dispatch(archivePlaylistItem(ownProps.item))
    },
    changeTrack: ((item: PlaylistItemType) => {
      dispatch(changeTrack(item))
    })
  }
}

export const mapStateToProps = (newState: any) => {
  return {
    audioPlayer: newState.audioPlayer
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(AudioPlayerPresenter)
