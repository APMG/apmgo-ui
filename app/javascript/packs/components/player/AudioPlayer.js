// @flow
import React from 'react';
import { connect, dispatch } from 'react-redux'

import AudioPlayerState from '../../models/AudioPlayerState';
import type { PlaylistItemType } from '../../redux/types';
import PlayPauseButton from './PlayPauseButton'
import MuteButton from './MuteButton'
import TimeScrubber from './TimeScrubber'
import TimeKeeper from './TimeKeeper'
import VolumeSlider from './VolumeSlider'
import PlayTimeDisplay from './PlayTimeDisplay'
import { audioMetaDataLoaded, audioCanPlay } from '../../redux/audio-player'

type AudioPlayerProps = {
  item: PlaylistItemType,
  playerState: AudioPlayerState,
  audioCanPlay: () => {},
  audioMetaDataLoaded: () => {}
}

export class AudioPlayerPresenter extends React.Component {

  audio: HTMLAudioElement
  props: AudioPlayerProps

  constructor(props: AudioPlayerProps) {
    super(props)
  }

  componentDidMount() {
    // by now, audio is an HTMLAudioElement ref
    this.audio.src = this.props.item.attributes.audio_url
  }

  componentWillReceiveProps(newProps: AudioPlayerProps) {
    this._setPlayPaused(newProps.playerState.paused)
    this.audio.muted = newProps.playerState.muted
    this.audio.volume = newProps.playerState.volume

    if(newProps.item.id !== this.props.item.id) {
      this.audio.src = newProps.item.attributes.audio_url
    }

    if(newProps.playerState.updateAudioElementTime) {
      this.audio.currentTime = newProps.playerState.currentTime
    }
  }

  _setPlayPaused(paused: boolean) {
    if (paused !== this.audio.paused) {
      if (paused) {
        this.audio.pause()
      } else {
        this.audio.play()
      }
    }
  }

  metaDataLoaded(a:any,b:any,c:any) {
    this.audio.currentTime = this.props.item.attributes.playtime
    this.props.audioMetaDataLoaded(this.audio.duration)
  }

  render() {
    return (
      <div>
        <h2>My Custom Playlist</h2>
        <h3>
            {this.props.item
              ? 'Now playing: ' + this.props.item.attributes.audio_title
              : 'Loading ...'}
        </h3>

        <audio
          ref={(ref) => this.audio = ref }
          onCanPlay={this.props.audioCanPlay}
          onLoadedMetadata={this.metaDataLoaded.bind(this)}>
        </audio>

        <PlayPauseButton />
        <MuteButton />

        <h3>Time Control</h3>
        <TimeKeeper // this component is invisible
          audio={ this.audio }
          item_id={this.props.item.id}
        />
        <PlayTimeDisplay currentTime={this.props.playerState.currentTime} />
        <TimeScrubber
          item_id={ this.props.item.id }
          currentTime={ this.props.playerState.currentTime }
          paused={ this.props.playerState.paused }
        />

        <h3>Volume Control</h3>
        <VolumeSlider />
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    audioCanPlay: () => {
      dispatch(audioCanPlay())
    },
    audioMetaDataLoaded: (duration) => {
      dispatch(audioMetaDataLoaded(duration))
    }
  }
}

const mapStateToProps = (newState) => {
  return {
    playerState: Object.assign({}, newState.audioPlayer)
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(AudioPlayerPresenter)
