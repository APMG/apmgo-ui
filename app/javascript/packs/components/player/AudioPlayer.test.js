import React from 'react'
import { AudioPlayerPresenter } from './AudioPlayer'
import getSnapshotJson, { itemFixtures } from '../../redux/__tests__/mock-initial-state'
import AudioPlayerState from '../../models/AudioPlayerState'

class TestAudioPlayerPresenter extends AudioPlayerPresenter {
  componentDidMount() {
    // noop
    // because the audio element ref that is set during render()
    // is not getting set by jest, and this method throws an error
  }
}

describe('AudioPlayer Component Test', () => {
  describe('Presenter', () => {
    it('Renders', () => {
      let tree = getSnapshotJson(<TestAudioPlayerPresenter item={ itemFixtures[0] } />)
      expect(tree).toMatchSnapshot();
    })
  })
})
