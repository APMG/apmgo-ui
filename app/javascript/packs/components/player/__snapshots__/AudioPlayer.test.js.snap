// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioPlayer Component Test Presenter Renders 1`] = `
<div>
  <h2>
    My Custom Playlist
  </h2>
  <h3>
    Now playing: Test Audio
  </h3>
  <audio
    onCanPlay={undefined}
    onLoadedMetadata={[Function]}
  />
  <button
    disabled={true}
    id="play-pause-loading-button"
  >
    Loading ...
  </button>
  <button
    id="mute-button"
    onClick={[Function]}
  >
    Mute
  </button>
  <h3>
    Time Control
  </h3>
  <div />
  <div>
    0:00
  </div>
  <div>
    <div
      style={
        Object {
          "muiPrepared": true,
        }
      }
      value={0}
    >
      <span />
      <span />
      <div
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "cursor": "default",
            "height": 18,
            "marginBottom": 48,
            "marginTop": 24,
            "muiPrepared": true,
            "position": "relative",
            "touchCallout": "none",
            "userSelect": "none",
            "width": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "height": 2,
              "left": 0,
              "muiPrepared": true,
              "position": "absolute",
              "top": 8,
              "width": "100%",
            }
          }
        >
          <div
            style={
              Object {
                "[object Object]": 0,
                "backgroundColor": "#00bcd4",
                "height": "100%",
                "left": 0,
                "marginRight": 6,
                "muiPrepared": true,
                "position": "absolute",
                "transition": "margin 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                "width": "calc(0%)",
              }
            }
          />
          <div
            style={
              Object {
                "[object Object]": 0,
                "backgroundColor": "#bdbdbd",
                "height": "100%",
                "marginLeft": 6,
                "muiPrepared": true,
                "position": "absolute",
                "right": 0,
                "transition": "margin 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                "width": "calc(100%)",
              }
            }
          />
          <div
            style={
              Object {
                "backgroundClip": "padding-box",
                "backgroundColor": "#ffffff",
                "border": "2px solid #bdbdbd",
                "borderRadius": "50%",
                "boxShadow": "none",
                "boxSizing": "border-box",
                "cursor": "pointer",
                "height": 12,
                "left": "0%",
                "margin": "1px 0 0 0",
                "muiPrepared": true,
                "outline": "none",
                "overflow": "visible",
                "pointerEvents": "inherit",
                "position": "absolute",
                "top": 0,
                "transform": "translate(-50%, -50%)",
                "transition": "background 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms, border-color 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms, width 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms, height 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                "width": 12,
                "zIndex": 1,
              }
            }
            tabIndex={0}
          >
            <div
              style={
                Object {
                  "height": 12,
                  "left": -2,
                  "muiPrepared": true,
                  "overflow": "visible",
                  "position": "absolute",
                  "top": -2,
                  "width": 12,
                }
              }
            />
          </div>
        </div>
      </div>
      <input
        max={1}
        min={0}
        name={undefined}
        required={true}
        step={0.01}
        type="hidden"
        value={0}
      />
    </div>
  </div>
  <h3>
    Volume Control
  </h3>
  <div
    style={
      Object {
        "muiPrepared": true,
      }
    }
    value={1}
  >
    <span />
    <span />
    <div
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "cursor": "default",
          "height": 18,
          "marginBottom": 48,
          "marginTop": 24,
          "muiPrepared": true,
          "position": "relative",
          "touchCallout": "none",
          "userSelect": "none",
          "width": "100%",
        }
      }
    >
      <div
        style={
          Object {
            "height": 2,
            "left": 0,
            "muiPrepared": true,
            "position": "absolute",
            "top": 8,
            "width": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "[object Object]": 0,
              "backgroundColor": "#00bcd4",
              "height": "100%",
              "left": 0,
              "marginRight": 6,
              "muiPrepared": true,
              "position": "absolute",
              "transition": "margin 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              "width": "calc(100%)",
            }
          }
        />
        <div
          style={
            Object {
              "[object Object]": 0,
              "backgroundColor": "#bdbdbd",
              "height": "100%",
              "marginLeft": 6,
              "muiPrepared": true,
              "position": "absolute",
              "right": 0,
              "transition": "margin 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              "width": "calc(0%)",
            }
          }
        />
        <div
          style={
            Object {
              "backgroundClip": "padding-box",
              "backgroundColor": "#00bcd4",
              "border": "0px solid transparent",
              "borderRadius": "50%",
              "boxSizing": "border-box",
              "cursor": "pointer",
              "height": 12,
              "left": "100%",
              "margin": "1px 0 0 0",
              "muiPrepared": true,
              "outline": "none",
              "overflow": "visible",
              "pointerEvents": "inherit",
              "position": "absolute",
              "top": 0,
              "transform": "translate(-50%, -50%)",
              "transition": "background 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms, border-color 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms, width 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms, height 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              "width": 12,
              "zIndex": 1,
            }
          }
          tabIndex={0}
        >
          <div
            style={
              Object {
                "height": 12,
                "left": 0,
                "muiPrepared": true,
                "overflow": "visible",
                "position": "absolute",
                "top": 0,
                "width": 12,
              }
            }
          />
        </div>
      </div>
    </div>
    <input
      max={1}
      min={0}
      name={undefined}
      required={true}
      step={0.01}
      type="hidden"
      value={1}
    />
  </div>
</div>
`;
